source/s3_auto_ranged_get.c

```diff
@@ -677,21 +787,34 @@ static void s_s3_auto_ranged_get_request_finished(
         /* If the object range was found, then record it. */
         if (found_object_size) {
             AWS_ASSERT(!auto_ranged_get->synced_data.object_range_known);
-
             auto_ranged_get->synced_data.object_range_known = true;
             auto_ranged_get->synced_data.object_range_empty = (total_content_length == 0);
             auto_ranged_get->synced_data.object_range_start = object_range_start;
             auto_ranged_get->synced_data.object_range_end = object_range_end;
-            auto_ranged_get->synced_data.total_num_parts =
-                aws_s3_get_num_parts(meta_request->part_size, object_range_start, object_range_end);
+            if (!first_part_size_mismatch) {
+                auto_ranged_get->synced_data.first_part_size = first_part_size;
+            }
+            auto_ranged_get->synced_data.total_num_parts = aws_s3_calculate_auto_ranged_get_num_parts(
+                meta_request->part_size,
+                auto_ranged_get->synced_data.first_part_size,
+                object_range_start,
+                object_range_end);
         }
 
         switch (request->request_tag) {
             case AWS_S3_AUTO_RANGE_GET_REQUEST_TYPE_HEAD_OBJECT:
                 auto_ranged_get->synced_data.head_object_completed = true;
                 AWS_LOGF_DEBUG(AWS_LS_S3_META_REQUEST, "id=%p Head object completed.", (void *)meta_request);
                 break;
-            case AWS_S3_AUTO_RANGE_GET_REQUEST_TYPE_PART:
+            case AWS_S3_AUTO_RANGE_GET_REQUEST_TYPE_GET_PART_NUMBER:
+                AWS_LOGF_DEBUG(AWS_LS_S3_META_REQUEST, "id=%p Get Part Number completed.", (void *)meta_request);
+                if (first_part_size_mismatch && found_object_size) {
+                    /* Try to fetch the first part again as a ranged get */
```

Comment:
trivial:
```suggestion
                    /* We canceled GET_PART_NUMBER_1 because it was bigger than expected.
                     * Try to fetch the first part again as a ranged get */
```